### DISCUSSION TITLE
Any LWFA demo input deck ?

### DISCUSSION THREAD

**Discussion by DanRRRR** [epochpic/epoch#603] [General]
Do code users or developers have any relatively simple and fast 2D example with LWFA and which of course demonstrates gain of energy with time?

**Comment by DanRRRR** [comment]
I found several nice descriptions and examples here [https://epochpic.github.io/quickstart/basic_examples/particle_load.html#load-particles-from-file](url) Is this new recent stuff? That is what manual needs more, more examples, more practical details

Here are related questions about loading particles from file, seems needed for LWFA

1) Is it possible to add particles at specific time after run starts, not at the start of simulation? Things are that many simulations including  LWFA  have to delay laser peak 3 and better 4 pulsewidths before time zero. During this time the initial particles will go far ahead of formed bubble. may it is possible to setup the stream of electrons in time and space same way like the laser is set?

2) Example was done with MATLAB. Of course this is much better than nothing but  MATLAB is just one of few dozen languages. Things are that it is absolutely impossible to convince people to change their preferred compiler, editor (even if it is dumbest ever VI versus greatest ever Kate) and political views :)
So the following question obviously is: how these files preparation will look in other languages? Like Fortran, Python, C or just generally what is the structure of such files made with MATLAB?

3) And even better, how about usual ASCII files for that? Sometimes it is necessary  to add just few particles which would be easy to be done just in usual ASCII Notepad. Can be such input files be made  in regular text editor? Or they have to be binary (unformatted in Fortran),  do they also have to be with STREAM attribute? 

   If ASCII not possible it would be great to add this functionality, i think. Esier to debug. Then we would easily turn the file to binary for speed.

4) How about adding more practical momentum unit  mc instead of using SI for it ? Users will set the value for momentum in clearly understandable units like this: drift_x_mc = 1 instead of drft_x = 2.898765432e-22  :)  . Sorry if this already exists

**Comment by Status-Mirror** [comment]
Hey Dan,

I don't have a small-scale full-LWFA example for you, but hopefully the demos should be enough to point you in the right direction. Regarding your specific questions:

1. To add specific macro-particles into the simulation window at different start times, you may be interested in using the [file-injectors](https://epochpic.github.io/quickstart/basic_examples/injectors.html#file-injectors). Here you can specify the time, weight and momentum of each new particle, but this only works for particles injected from the boundary of the simulation window. You'll be pleased to know that file-injectors only support ASCII files.

2. The structure of the binary file for the particle-loader is simply a list of double-precision numbers. Most languages should offer a way to write double-precision numbers in binary format. I simply chose MATLAB for the example as this is what I'm most comfortable with.

3. Unfortunately we can't offer ASCII support for particle-loading. This would require a modification to the maths parser, and we will no longer be supporting large developments to the FORTRAN code, only critical bug-fixes. The team's development time is to be instead spent on the C++ upgrade.

4. This doesn't exist, but you could achieve something similar using the built-in [maths parser constants](https://epochpic.github.io/documentation/code_details/maths_parser.html#constants). Something like:

```
  drift_x = 1.0 * me * c
```

Cheers,
Stuart

**Comment by DanRRRR** [comment]
Thanks Stuart,
1) I was confused in the initial documentation, in these examples in this link above and still in your respond about ASCII vs binary which you use simultaneously. By ASCII by definition is meant part of 8 byte word which is human readable like the text you see, by binary -- all 8 bytes of the word containing special characters, including not human readable. To stop this confusion, if it is ASCII like you are mentioning can you post here initial part of any file mentioned in the demo?
Things also are that if this is ASCII then for example some languages have limit on number of characters in the line while files could be 1000x larger than that. Hence not clear to me how to write these files, how many symbols/bytes per one line ? Hence i asked  to post the example on different languages to be clear here. Matlab files are like a black hole. Even if they are ASCII human readable they still contain special binary codes at the start and the end of file and at the end of each record. So essentially they are proprietary and not easily portable. The only portable files i personally assessed with different compilers are pure ASCII txt and cvs files and binary with 'STREAM' attribute. 

2) You probably have noticed a lot of people are interested with LWFA recently. This is because a new possibility was opened by boosted frame method which may increase speed 100 and 1000 times for relativistic cases, and all of them are relativistic in some extent. Even more, if boosted frame is not supported then any usual LWFA with moving window becomes almost useless for the applications which are in development, you can not simulate 3-10 meters long accelerator with it.  The potential users of EPOCH may abandon it to other open source PIC codes (you may see huge boost in activity there) which support it while EPOCH is in rebuild stage. Can you please look at these and other tricks with LWFA more closely?  

3) Posting examples is essential to keep people with this code, with any code. Users still can practice with the smaller cases and easily pass learning curve. I for example lost 1/2 years on LWFA because of the EPOCH bug in some specific settings and absence of examples. Tried many of them and only one finally (partially) started to behave recently, but it still looks too artificial

4) Are you also using ChatGPT to transfer EPOCH to C code ? 

/* Then after your rebuild and hopefully improve it ChatGPT will transfer it back to Fortran in seconds  :) :) Somebody will probably transfer it from Fortran 2003 to Fortran-95 or even 77, which will make it even faster, not slower, and more user-friendly,  because decently the way it is currently written is sometimes too complex, pointers. pointer, linked lists, proprietary files...It's essentially a "single actor theater" code

**Reply by Status-Mirror** [reply]
1. Only file-injectors support ASCII files, and the contents of these files are given in the [file-injectors demo](https://epochpic.github.io/quickstart/basic_examples/injectors.html#file-injectors). The reason for this weirdness is simply that I wrote the file-injectors module during my PhD, and I found ASCII easier to work with than binary. The rest of the code was written long before me, and was set up to run binary. Binary files are simply a line of double-precision binary numbers, with nothing between them - no special weirdness is introduced by writing them with MATLAB.
2. Unfortunately I can't commit to working on boosted frames at the moment, developer time is spent in too many other places. Other people are welcome to try adding this capability themselves, or they may use one of the other codes you refer to.
3. I'll slowly continue my trickle of demo-writing. It will be useful for users, but also for test cases in EPOC++.
4. I think EPOCH may be too big for ChatGPT, and I don't think this would be desirable anyway. Our developers are doing lots of tests to optimise epoch for C++ compilers, and there are new features we want to support, so it won't be a straightforward conversion.
